### ISSUE TITLE
SSLHandshakeException for some TLS version starting in 4.2.0

### ISSUE THREAD

**Issue by ParkerK** [bumptech/glide#2832]
After upgrading to Glide 4.2+, I am having issues loading some HTTPS images from a specific server, example url is: https://lainchan.org/diy/thumb/1506629821713.png 

It works flawlessly in Glide 4.1.1 and lower, so I am not sure if this is Glide specific or OKHttp (if the version Glide uses was changed in 4.2.0). This issue is also only for certain URLs, it still loads most HTTPs urls without issue. 

Cached images still load, but new calls fail. I am guessing this is an OkHttp3 issue, was the version used changed when Glide bumped from 4.1.1 to 4.2.0?

**Glide Version**:
Issue occurs on 4.2.0 and up

**Integration libraries**:
OkHttp3, using `com.github.bumptech.glide:recyclerview-integration:4.2.0` and up

**Device/Android Version**:
Issue occurs on Samsung Galaxy S8, will try on other versions on emulator

**Issue details / Repro steps / Use case background**: 
I was able to fix this, by manually setting the TLS encodings in OkHTTP

**Stack trace / LogCat**:
```class com.bumptech.glide.load.engine.GlideException: Failed to load resource
                     Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetching data failed, class java.io.InputStream, REMOTE
                       Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetch failed
                         Cause (1 of 1): class com.bumptech.glide.load.engine.GlideException: Fetch failed
                           Cause (1 of 1): class javax.net.ssl.SSLHandshakeException: Handshake failed
com.app.debug I/Glide: Root cause (1 of 1)
                   javax.net.ssl.SSLHandshakeException: Handshake failed
                       at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:444)
                       at okhttp3.internal.connection.RealConnection.connectTls(RealConnection.java:299)
                       at okhttp3.internal.connection.RealConnection.establishProtocol(RealConnection.java:268)
                       at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:160)
                       at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:256)
                       at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:134)
                       at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:113)
                       at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
                       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
                       at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
                       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
                       at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
                       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                       at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125)
                       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                       at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
                       at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
                       at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
                       at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
                       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                       at java.lang.Thread.run(Thread.java:762)
                    Caused by: javax.net.ssl.SSLProtocolException: SSL handshake terminated: ssl=0x7f8abd5400: Failure in SSL library, usually a protocol error
                   error:10000410:SSL routines:OPENSSL_internal:SSLV3_ALERT_HANDSHAKE_FAILURE (external/boringssl/src/ssl/s3_pkt.c:641 0x7f67a0c5a0:0x00000001)
                   error:1000009a:SSL routines:OPENSSL_internal:HANDSHAKE_FAILURE_ON_CLIENT_HELLO (external/boringssl/src/ssl/s3_clnt.c:800 0x7f890a5467:0x00000000)
                       at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
                       at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:364)
                       	... 23 more
```

**Comment by ParkerK** [comment]
To try to debug, is there a way I can force Glide to use a different version of OkHttps3, so I can try to pinpoint if it's the issue, and if so, when exactly it broke?

**Comment by sjudd** [comment]
You can exclude the okhttp dependency from Glide's okhttp integration library and then specify whatever version you want in your build.gradle file:

```groovy
implementation ("com.github.bumptech.glide.integration:okhttp:4.5.0") {
   exclude group: "com.squareup.okhttp:"
}
// Then whatever version you want here:
implementation "com.squareup.okhttp:okhttp:2.7.5"
```

You can also test by just loading the image with OkHttp directly if this is reliably reproducible.

**Comment by sjudd** [comment]
Hi @ParkerK,

It looks like you might not have included some important information when you filed this issue. To make sure we have all the information we need to help with your issue. Please take a look at the [Issue Template](https://github.com/bumptech/glide/blob/master/ISSUE_TEMPLATE.md) and make sure you've filled out as many sections as possible.

In particular you **must** include:

* Glide Version
* Glide load line
* Issue details.

I'm marking this issue as requiring more information. Please understand that if you do not provide that information or update the issue within the next 7 days this issue will be automatically closed so that it doesn''t clutter the bug tracker.

Best regards,

~ Your friendly GitIssueBot

PS: I'm just an automated script, not a human being.

**Comment by ParkerK** [comment]
Thanks, doing this and specifying an older OkHttp3, but a newer Glide resulted in no issues. Which means it's purely an OkHttp3 issue, so I'll report this to them.

**Comment by sjudd** [comment]
Thanks for the follow up!

**Comment by Sdghasemi** [comment]
Hello @ParkerK,
Which versions are you using for Glide and OkHttp3 exactly? Experiencing the same problem here :(
It would be a great help if you also add your dependencies code here.
Thanks in advance.

**Comment by KeyStarr** [comment]
This issue arises due to the Android API v. 16-19 having TSL v1.2 protocol bundled but not enabled by default. So the device is trying to connect to the server which accepts TSL v1.2 only, fails and retreats to using SSL v.3.0 which also fails.

In order to fix this you need to enable TSL v1.2 on API 16-19 in code, a working solution can be found here: https://gist.github.com/kotomisak/e22f0469342319570770e928eaa3b8a2

**Comment by mahdit83** [comment]
> Thanks, doing this and specifying an older OkHttp3, but a newer Glide resulted in no issues. Which means it's purely an OkHttp3 issue, so I'll report this to them.

So is ti solved? I have same issue with last version of glide.

**Comment by mahdit83** [comment]
> This issue arises due to the Android API v. 16-19 having TSL v1.2 protocol bundled but not enabled by default. So the device is trying to connect to the server which accepts TSL v1.2 only, fails and then retreats to using SSL v.3.0 which also fails.
> 
> In order to fix this you need to enable TSL v1.2 on API 16-19 in code, a working solution can be found here: https://gist.github.com/kotomisak/e22f0469342319570770e928eaa3b8a2

@KeyStarr Your link is broken.

**Comment by mahdit83** [comment]
I got this problem in Emulator Api 29.

**Comment by superdiazzz** [comment]
> This issue arises due to the Android API v. 16-19 having TSL v1.2 protocol bundled but not enabled by default. So the device is trying to connect to the server which accepts TSL v1.2 only, fails and then retreats to using SSL v.3.0 which also fails.
> 
> In order to fix this you need to enable TSL v1.2 on API 16-19 in code, a working solution can be found here: https://gist.github.com/kotomisak/e22f0469342319570770e928eaa3b8a2

url is 404

**Comment by superdiazzz** [comment]
> This issue arises due to the Android API v. 16-19 having TSL v1.2 protocol bundled but not enabled by default. So the device is trying to connect to the server which accepts TSL v1.2 only, fails and then retreats to using SSL v.3.0 which also fails.
> 
> In order to fix this you need to enable TSL v1.2 on API 16-19 in code, a working solution can be found here: https://gist.github.com/kotomisak/e22f0469342319570770e928eaa3b8a2

@KeyStarr  Url is 404
